pkg/terraform/workspace.go

```diff
@@ -0,0 +1,262 @@
+	LastOperationError error
+}
+
+func (w *Workspace) Refresh(ctx context.Context) (RefreshResult, error) {
+	if w.LastOperation.StartTime != nil {
+		// The last operation is still ongoing.
+		if w.LastOperation.EndTime == nil {
+			return RefreshResult{
+				IsApplying:   w.LastOperation.Type == "apply",
+				IsDestroying: w.LastOperation.Type == "destroy",
+			}, nil
+		}
+		// We know that the operation finished, so we need to flush so that new
+		// operation can be started.
+		defer w.LastOperation.Flush()
+
+		// The last operation finished with error.
+		if w.LastOperation.Err != nil {
+			return RefreshResult{
+				IsApplying:         w.LastOperation.Type == "apply",
+				IsDestroying:       w.LastOperation.Type == "destroy",
+				LastOperationError: errors.Wrapf(w.LastOperation.Err, "%s operation failed", w.LastOperation.Type),
+			}, nil
+		}
+		// The deletion is completed so there is no resource to refresh.
+		if w.LastOperation.Type == "destroy" {
+			return RefreshResult{}, kerrors.NewNotFound(schema.GroupResource{}, "")
+		}
+	}
+	cmd := exec.CommandContext(ctx, "terraform", "apply", "-refresh-only", "-auto-approve", "-input=false", "-json")
+	cmd.Dir = w.dir
+	out, err := cmd.CombinedOutput()
+	w.logger.Debug("refresh completed", "out", string(out))
+	if err != nil {
+		return RefreshResult{}, errors.Wrapf(err, "cannot refresh: %s", string(out))
+	}
+	raw, err := os.ReadFile(filepath.Join(w.dir, "terraform.tfstate"))
+	if err != nil {
+		return RefreshResult{}, errors.Wrap(err, "cannot read terraform state file")
+	}
+	s := &json.StateV4{}
+	if err := json.JSParser.Unmarshal(raw, s); err != nil {
+		return RefreshResult{}, errors.Wrap(err, "cannot unmarshal tfstate file")
+	}
+	if len(s.Resources) == 0 {
+		return RefreshResult{}, kerrors.NewNotFound(schema.GroupResource{}, "")
+	}
+	return RefreshResult{State: s}, nil
+}
+
+type PlanResult struct {
+	Exists   bool
+	UpToDate bool
+}
+
+func (w *Workspace) Plan(ctx context.Context) (PlanResult, error) {
+	// The last operation is still ongoing.
+	if w.LastOperation.StartTime != nil && w.LastOperation.EndTime == nil {
+		return PlanResult{}, errors.Errorf("%s operation that started at %s is still running", w.LastOperation.Type, w.LastOperation.StartTime.String())
+	}
+	cmd := exec.CommandContext(ctx, "terraform", "plan", "-refresh=false", "-input=false", "-json")
+	cmd.Dir = w.dir
+	out, err := cmd.CombinedOutput()
+	w.logger.Debug("plan completed", "out", string(out))
+	if err != nil {
+		return PlanResult{}, errors.Wrapf(err, "cannot plan: %s", string(out))
+	}
+	line := ""
+	for _, l := range strings.Split(string(out), "\n") {
+		if strings.Contains(l, `"type":"change_summary"`) {
+			line = l
+			break
+		}
+	}
+	if line == "" {
+		return PlanResult{}, errors.Errorf("cannot find the change summary line in plan log: %s", string(out))
+	}
+	type plan struct {
+		Changes struct {
+			Add    float64 `json:"add,omitempty"`
+			Change float64 `json:"change,omitempty"`
```

Comment:
Could we also return a Diff here on what needs to be updated, which could finally set as[ExternalObservation.Diff](https://github.com/crossplane/crossplane-runtime/blob/d5661210733b2bf926281f231ea1db9e5be7bab9/pkg/reconciler/managed/reconciler.go#L336)?

This would be quite helpful for debugging infamous getting constant updates we observe in other providers.

Reply:
Added https://github.com/crossplane-contrib/terrajet/issues/44#issuecomment-935012304 to cover both diff and error reporting, assuming the JSON parsing would do both.